File: android/localasset-core/src/main/kotlin/io/github/xesam/android/localasset/core/engine/DefaultLocalAssetEngine.kt
package io.github.xesam.android.localasset.core.engine

import io.github.xesam.android.localasset.core.api.EngineObserver
import io.github.xesam.android.localasset.core.api.EngineResult
import io.github.xesam.android.localasset.core.api.EngineStageEvent
import io.github.xesam.android.localasset.core.api.LocalAssetEngine
import io.github.xesam.android.localasset.core.api.Policy
import io.github.xesam.android.localasset.core.api.ResolverResult
import io.github.xesam.android.localasset.core.api.ResourceLoader
import io.github.xesam.android.localasset.core.api.SchemeAdapter
import io.github.xesam.android.localasset.core.error.ResourceErrorCategory
import io.github.xesam.android.localasset.core.error.ResourceException
import io.github.xesam.android.localasset.core.internal.AdapterRegistry
import io.github.xesam.android.localasset.core.internal.ResolverChain
import io.github.xesam.android.localasset.core.model.AssetRequest
import io.github.xesam.android.localasset.core.model.ResolveContext
import io.github.xesam.android.localasset.core.model.ResourceDescriptor

class DefaultLocalAssetEngine(
    adapters: List<SchemeAdapter>,
    private val resolverChain: ResolverChain,
    private val policy: Policy,
    private val loader: ResourceLoader,
    private val observer: EngineObserver? = null,
) : LocalAssetEngine {
    private val adapterRegistry = AdapterRegistry(adapters)

    override fun resolve(url: String, context: ResolveContext): EngineResult {
        val request: AssetRequest
        try {
            request = adapterRegistry.select(url).parse(url)
        } catch (e: Exception) {
            return stageFailure(e, EngineObserver.STAGE_ADAPTER_PARSE, request = null,
                defaultCategory = ResourceErrorCategory.PARSE_ERROR, defaultPrefix = "invalid url: ")
        }
        emit(EngineObserver.STAGE_ADAPTER_PARSE, request = request)
        return resolve(request, context)
    }

    override fun resolve(request: AssetRequest, context: ResolveContext): EngineResult {
        try {
            policy.preCheck(request, context)
        } catch (e: Exception) {
            return stageFailure(e, EngineObserver.STAGE_POLICY_PRE, request,
                defaultCategory = ResourceErrorCategory.SECURITY_ERROR, defaultPrefix = "preCheck failed: ")
        }
        emit(EngineObserver.STAGE_POLICY_PRE, request = request)

        val result: ResolverResult
        try {
            result = resolverChain.resolve(request, context, observer)
        } catch (e: Exception) {
            return stageFailure(e, EngineObserver.STAGE_RESOLVE, request,
                defaultCategory = ResourceErrorCategory.RESOLUTION_ERROR, defaultPrefix = "resolver chain failed: ")
        }

        return when (result) {
            is ResolverResult.Hit -> resolveHit(request, result.descriptor, context)
            is ResolverResult.Failure -> {
                val failure = EngineResult.Failure(result.category, result.reason, EngineObserver.STAGE_RESOLVE)
                emit(EngineObserver.STAGE_RESOLVE, request = request, failure = failure)
                failure
            }
            ResolverResult.Skip -> {
                val failure = EngineResult.Failure(
                    ResourceErrorCategory.RESOLUTION_ERROR,
                    "resource not found",
                    EngineObserver.STAGE_RESOLVE,
                )
                emit(EngineObserver.STAGE_RESOLVE, request = request, failure = failure)
                failure
            }
        }
    }

    private fun resolveHit(
        request: AssetRequest,
        descriptor: ResourceDescriptor,
        context: ResolveContext,
    ): EngineResult {
        try {
            policy.postCheck(request, descriptor, context)
        } catch (e: Exception) {
            return stageFailure(e, EngineObserver.STAGE_POLICY_POST, request, descriptor,
                defaultCategory = ResourceErrorCategory.SECURITY_ERROR, defaultPrefix = "postCheck failed: ")
        }
        emit(EngineObserver.STAGE_POLICY_POST, request = request, descriptor = descriptor)

        try {
            val data = loader.load(descriptor)
            emit(EngineObserver.STAGE_LOAD, request = request, descriptor = descriptor)
            emit(EngineObserver.STAGE_COMPLETE, request = request, descriptor = descriptor)
            return EngineResult.Success(descriptor, data)
        } catch (e: Exception) {
            return stageFailure(e, EngineObserver.STAGE_LOAD, request, descriptor,
                defaultCategory = ResourceErrorCategory.LOAD_ERROR, defaultPrefix = "unexpected load failure: ")
        }
    }

    /**
     * Unified error handler for all pipeline stages: extracts [ResourceException] category/stage
     * when present, otherwise wraps with a stage-specific fallback category and prefix; emits the
     * failure event and returns the [EngineResult.Failure].
     *
     * Mirrors the ArkTS `toFailure` + emit pattern and the Swift `emitFailure` pattern, keeping
     * error messages and stage assignments identical across platforms.
     */
    private fun stageFailure(
        error: Exception,
        stage: String,
        request: AssetRequest?,
        descriptor: ResourceDescriptor? = null,
        defaultCategory: ResourceErrorCategory,
        defaultPrefix: String,
    ): EngineResult.Failure {
        val failure = if (error is ResourceException) {
            EngineResult.Failure(error.category, error.message, error.stage ?: stage)
        } else {
            EngineResult.Failure(defaultCategory, "$defaultPrefix${error.message}", stage)
        }
        emit(stage, request, descriptor, failure)
        return failure
    }

    private fun emit(
        stage: String,
        request: AssetRequest?,
        descriptor: ResourceDescriptor? = null,
        failure: EngineResult.Failure? = null,
    ) {
        observer?.onStage(
            EngineStageEvent(
                stage = stage,
                request = request,
                descriptor = descriptor,
                failure = failure,
            ),
        )
    }
}

File: ios/localasset-core/Sources/LocalAssetCore/engine/DefaultLocalAssetEngine.swift
public class DefaultLocalAssetEngine: LocalAssetEngine {
    private let adapterRegistry: AdapterRegistry
    private let resolverChain: ResolverChain
    private let policy: Policy
    private let loader: ResourceLoader
    private let observer: EngineObserver?

    public init(
        adapters: [SchemeAdapter],
        resolverChain: ResolverChain,
        policy: Policy,
        loader: ResourceLoader,
        observer: EngineObserver? = nil
    ) {
        self.adapterRegistry = AdapterRegistry(adapters: adapters)
        self.resolverChain = resolverChain
        self.policy = policy
        self.loader = loader
        self.observer = observer
    }

    public func resolve(url: String, context: ResolveContext) -> EngineResult {
        let request: AssetRequest
        do {
            let adapter = try adapterRegistry.select(url: url)
            request = try adapter.parse(url: url)
        } catch {
            return stageFailure(error, stage: EngineStage.adapterParse, request: nil,
                defaultCategory: .parseError, defaultPrefix: "invalid url: ")
        }
        emit(EngineStage.adapterParse, request: request)
        return resolve(request: request, context: context)
    }

    public func resolve(request: AssetRequest, context: ResolveContext) -> EngineResult {
        do {
            try policy.preCheck(request: request, context: context)
        } catch {
            return stageFailure(error, stage: EngineStage.policyPre, request: request,
                defaultCategory: .securityError, defaultPrefix: "preCheck failed: ")
        }
        emit(EngineStage.policyPre, request: request)

        let resolverResult = resolverChain.resolve(request: request, context: context, observer: observer)

        switch resolverResult {
        case .skip:
            return emitFailure(
                EngineStage.resolve, request: request,
                failure: EngineFailure(category: .resolutionError, reason: "resource not found", stage: EngineStage.resolve)
            )
        case .failure(let category, let reason):
            return emitFailure(
                EngineStage.resolve, request: request,
                failure: EngineFailure(category: category, reason: reason, stage: EngineStage.resolve)
            )
        case .hit(let descriptor):
            return resolveHit(request: request, descriptor: descriptor, context: context)
        }
    }

    private func resolveHit(
        request: AssetRequest,
        descriptor: ResourceDescriptor,
        context: ResolveContext
    ) -> EngineResult {
        do {
            try policy.postCheck(request: request, descriptor: descriptor, context: context)
        } catch {
            return stageFailure(error, stage: EngineStage.policyPost, request: request, descriptor: descriptor,
                defaultCategory: .securityError, defaultPrefix: "postCheck failed: ")
        }
        emit(EngineStage.policyPost, request: request, descriptor: descriptor)

        do {
            let data = try loader.load(descriptor: descriptor)
            emit(EngineStage.load, request: request, descriptor: descriptor)
            emit(EngineStage.complete, request: request, descriptor: descriptor)
            return .success(descriptor: descriptor, data: data)
        } catch {
            return stageFailure(error, stage: EngineStage.load, request: request, descriptor: descriptor,
                defaultCategory: .loadError, defaultPrefix: "unexpected load failure: ")
        }
    }

    // MARK: - Error handling helpers

    /// Unified error handler for throwing pipeline stages: extracts `ResourceException`
    /// category/stage when present, otherwise wraps with a stage-specific fallback category and
    /// prefix; emits the failure event and returns the matching `EngineResult.failure`.
    ///
    /// Mirrors the Kotlin `stageFailure` and ArkTS `toFailure` + emit pattern, keeping error
    /// messages and stage assignments identical across platforms.
    private func stageFailure(
        _ error: Error,
        stage: String,
        request: AssetRequest?,
        descriptor: ResourceDescriptor? = nil,
        defaultCategory: ResourceErrorCategory,
        defaultPrefix: String
    ) -> EngineResult {
        let failure: EngineFailure
        if let e = error as? ResourceException {
            failure = EngineFailure(category: e.category, reason: e.message, stage: e.stage ?? stage)
        } else {
            failure = EngineFailure(category: defaultCategory, reason: "\(defaultPrefix)\(error)", stage: stage)
        }
        emit(stage, request: request, descriptor: descriptor, failure: failure)
        return .failure(category: failure.category, reason: failure.reason, stage: failure.stage)
    }

    /// Emits `failure` to the observer (if attached) and returns the matching `EngineResult.failure`
    /// — used for non-throwing failure paths (ResolverResult.Failure / Skip).
    @discardableResult
    private func emitFailure(
        _ stage: String,
        request: AssetRequest?,
        descriptor: ResourceDescriptor? = nil,
        failure: EngineFailure
    ) -> EngineResult {
        emit(stage, request: request, descriptor: descriptor, failure: failure)
        return .failure(category: failure.category, reason: failure.reason, stage: failure.stage)
    }

    private func emit(
        _ stage: String,
        request: AssetRequest?,
        descriptor: ResourceDescriptor? = nil,
        failure: EngineFailure? = nil
    ) {
        observer?.onStage(EngineStageEvent(
            stage: stage,
            request: request,
            descriptor: descriptor,
            failure: failure
        ))
    }
}
